src/python-common/ceph/fs/earmarking.py

```diff
@@ -56,27 +65,63 @@ def _handle_cephfs_error(self, e: Exception, action: str) -> None:
             raise EarmarkException(-e.errno, e.strerror) from e
         else:
             log.error(f"Unexpected error {action} earmark: {e}")
-            raise EarmarkException(errno.EIO, "Unexpected error") from e
+            raise EarmarkException
+
+    @staticmethod
+    def parse_earmark(value: str) -> Optional[EarmarkContents]:
+        """
+        Parse an earmark value. Returns None if the value is an empty string.
+        Raises EarmarkParseError if the top-level scope is not valid or the earmark
+        string is not properly structured.
+        Returns an EarmarkContents for valid earmark values.
+
+        :param value: The earmark string to parse.
+        :return: An EarmarkContents instance if valid, None if empty.
+        """
+        if not value:
+            return None
+
+        parts = value.split('.')
+
+        # Check if the top-level scope is valid
+        if parts[0] not in (scope.value for scope in EarmarkTopScope):
+            raise EarmarkParseError(f"Invalid top-level scope: {parts[0]}")
+
+        # Check if all parts are non-empty to ensure valid dot-separated format
+        if not all(parts):
+            raise EarmarkParseError("Earmark contains empty sections.")
+
+        # Return parsed earmark with top scope and subsections
+        return EarmarkContents(top=EarmarkTopScope(parts[0]), subsections=parts[1:])
 
     def _validate_earmark(self, earmark: str) -> bool:
         """
-        Validates that the earmark string is either empty or composed of parts separated by scopes,
-        with the top-level scope being either 'nfs' or 'smb'.
+        Validates the earmark string further by checking specific conditions for scopes like 'smb'.
 
         :param earmark: The earmark string to validate.
         :return: True if valid, False otherwise.
         """
-        if not earmark or earmark in (scope.value for scope in EarmarkTopScope):
-            return True
-
-        parts = earmark.split('.')
+        try:
+            parsed = self.parse_earmark(earmark)
+
+            # If parsed is None, it's considered valid since the earmark is empty
+            if not parsed:
+                return True
+
+            # Specific validation for 'smb' scope
+            if parsed.top == EarmarkTopScope.SMB:
+                # Valid formats: 'smb' or 'smb.cluster.{cluster_id}'
+                if not (len(parsed.subsections) == 0 or
+                        (len(parsed.subsections) == 2 and
+                         parsed.subsections[0] == 'cluster' and parsed.subsections[1])):
+                    raise EarmarkParseError(
+                        "Invalid 'smb' earmark format. Must be 'smb' or 'smb.cluster.{cluster_id}'."
+                    )
 
-        if parts[0] not in (scope.value for scope in EarmarkTopScope):
+            return True
+        except EarmarkParseError:
             return False
```

Comment:
I'd rather see this moved up to just around `self.parse_earmark` and then you can unindent a bunch of logic.

Reply:
Done!